File: sdk/typescript/rooch-sdk/src/bcs/bcs.test.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { describe, it, expect } from 'vitest'

import { Ed25519Keypair, Secp256k1Keypair } from '../keypairs/index.js'
import { bytesEqual } from '../utils/index.js'

import { bcs } from './index.js'

describe('BCS', () => {
  it('Address', () => {
    const roochAddress = new Ed25519Keypair().getRoochAddress()

    const bcs1 = bcs.Address.serialize(roochAddress).toBytes()
    const bcs2 = bcs.Address.serialize(roochAddress.toHexAddress()).toBytes()
    const bcs3 = bcs.Address.serialize(roochAddress.toBech32Address()).toBytes()

    expect(bytesEqual(bcs1, bcs2) && bytesEqual(bcs1, bcs3)).toBeTruthy()

    const bitcoinAddress = new Secp256k1Keypair().getBitcoinAddress()

    const bcs4 = bcs.Address.serialize(bitcoinAddress).toBytes()
    const bcs5 = bcs.Address.serialize(bitcoinAddress.genRoochAddress()).toBytes()

    expect(bytesEqual(bcs4, bcs5)).toBeTruthy()
  })
})

File: sdk/golang/rooch-sdk/bcs/bcs_test.go
package bcs

import (
	"testing"

	"github.com/rooch-network/rooch/sdk/golang/keypairs"
	"github.com/rooch-network/rooch/sdk/golang/utils"
	"github.com/stretchr/testify/assert"
)

func TestBCS(t *testing.T) {
	t.Run("Address", func(t *testing.T) {
		// Test Ed25519 address serialization
		ed25519Keypair := keypairs.NewEd25519Keypair()
		roochAddress := ed25519Keypair.GetRoochAddress()

		bcs1 := BCS.Address.Serialize(roochAddress).ToBytes()
		bcs2 := BCS.Address.Serialize(roochAddress.ToHexAddress()).ToBytes()
		bcs3 := BCS.Address.Serialize(roochAddress.ToBech32Address()).ToBytes()

		assert.True(t, utils.BytesEqual(bcs1, bcs2))
		assert.True(t, utils.BytesEqual(bcs1, bcs3))

		// Test Secp256k1 address serialization
		secp256k1Keypair := keypairs.NewSecp256k1Keypair()
		bitcoinAddress := secp256k1Keypair.GetBitcoinAddress()

		bcs4 := BCS.Address.Serialize(bitcoinAddress).ToBytes()
		bcs5 := BCS.Address.Serialize(bitcoinAddress.GenRoochAddress()).ToBytes()

		assert.True(t, utils.BytesEqual(bcs4, bcs5))
	})
} 